### PR TITLE
enable plugins with transforms

### LINKED ISSUE #188: Plugins are not working
Thanks for creating this amazing package! This is a really awesome idea. 

I would like to ask for some help with plugins. For some reason the don't do anything.

My code
```
import React from "react";
import { render } from "react-dom";
const ReactMarkdown = require("react-markdown");
const toc = require("remark-toc");

const input = `
# Alpha

## Table of Contents

## Bravo

### Charlie

## Delta
`;

const App = () => <ReactMarkdown source={input} plugins={[toc]} />;

render(<App />, document.getElementById("root"));
```

codesandbox: https://codesandbox.io/s/340pq1r2l6

The output:

```
<div id="root">
    <div>
        <h1>Alpha</h1>
        <h2>Table of Contents</h2>
        <h2>Bravo</h2>
        <h3>Charlie</h3>
        <h2>Delta</h2>
    </div>
</div>
```

The table of contents is not created also slugs are not added to headings.

### PR DESCRIPTION
**Pull request by frankieali**
Allowing remark to execute `parser.runSync()` so that plugin transforms can be applied to the AST tree. Demo here: https://frankieali.github.io/react-markdown/demo/dist/

Closes #188

### CODE REVIEW

**Comment by frankieali** [comment]
I noticed that lists were not rendering as expected when I dropped [remark-toc](https://github.com/remarkjs/remark-toc) in for testing, since it was part of the original issue in #188. The renderer conditions were looking for `null` values but not `undefined` so I added them. It's not exactly related to `enabling plugins with transform` but it does relate to #188 so let me know if this should be separated out into a different PR.
I've updated my demo above with remark-TOC

**Comment by callain** [comment]
@frankieali Will this also works for plugin like remark-gemoji-to-emoji ? If this is the case I might need your patch 👍

**Comment by alifr** [comment]
@callain Yes, my demo worked fine when I dropped [remark-gemoji-to-emoji](https://github.com/jackycute/remark-gemoji-to-emoji) in.
I'm deep in another branch right now, but attached is a screen shot.
gemoji-to-emoji-example

**Comment by frankieali** [comment]
@callain Yes, my demo worked fine when I dropped [remark-gemoji-to-emoji](https://github.com/jackycute/remark-gemoji-to-emoji) in.
I'm deep in another branch right now, but attached is a screen shot.

gemoji-to-emoji-example

**Comment by callain** [comment]
@frankieali Cool ! 
@rexxars Do you have a moment to check this PR and make a version if everything is ok ?

File: src/ast-to-react.js [hunk 1]
```diff
@@ -2,12 +2,32 @@
...
+  if(node.type === 'root') {
+    root = node;
+  }
+
+  // nodes generated by plugins may not have position - setting to node parent's position
+  // TODO: there should be a way to get and assign a node's position after parse
+  if(node.position === undefined) {
+    visit(root, node, (nodeVisit, indexVisit, parentVisit) => {
+      // just in case parent does not have position either
+      const defaultPosition = {
```
Comment (rexxars) [inline src/ast-to-react.js]: Since defaultPosition is static, maybe hoist it to the top of the file?

File: src/ast-to-react.js [hunk 2]
```diff
@@ -2,12 +2,32 @@
...
 const React = require('react')
 const xtend = require('xtend')
+const visit = require('unist-util-visit')
+
+let root = '';
 
 function astToReact(node, options, parent = {}, index = 0) {
   const renderer = options.renderers[node.type]
 
+  if(node.type === 'root') {
```
Comment (rexxars) [inline src/ast-to-react.js]: Nitpick, but coding style wants a space between the if and the parentheses opening

File: src/ast-to-react.js [hunk 3]
```diff
@@ -47,7 +67,7 @@ function getNodeProps(node, key, opts, renderer, parent, index) {
   }
 
   if (opts.rawSourcePos && !isTagRenderer) {
-    props.sourcePosition = node.position
+    props.sourceposition = node.position
```
Comment (rexxars) [inline src/ast-to-react.js]: Was this accidental? Why was this changed?

**Comment by frankieali** [comment]
I've made the requested updates. React was complaining once that `sourcePosition` should be `sourceposition` on my local, but I didn't look into it enough to understand why. I've reverted it back.

As far as the tests go, I'm seeing one failure on rawSourcePos, but it seems more like a fix to me since the rawSourcePos snapshot is undefined and now the test is returning a position object.

```
should pass on raw source position to non-tag renderers if rawSourcePos option is enabled

    expect(received).toMatchSnapshot()

    Snapshot name: `should pass on raw source position to non-tag renderers if rawSourcePos option is enabled 1`

    - Snapshot
    + Received

    - undefined
    + Position {
    +   "end": Object {
    +     "column": 6,
    +     "line": 1,
    +     "offset": 5,
    +   },
    +   "indent": Array [],
    +   "start": Object {
    +     "column": 1,
    +     "line": 1,
    +     "offset": 0,
    +   },
    + }
```

**Comment by callain** [comment]
@rexxars Are the changes okay to you ?

**Comment by rexxars** [comment]
There are still no tests for the new functionality - adding a minor plugin to the test suite to ensure that the transformation step is actually is executed would be nice. I'm not sure if there are good reasons why the nodes without a position needs one - if they do, maybe add a comment as to why, and include some tests to ensure that the position gets assigned as expected?

`root` is also assigned in the global module scope - I think modifying the `astToReact` signature a bit would be better, ensuring that the root is always explicit (the root-level node passed from `react-markdown.js`). Also not sure why it's assigned to an empty string by default, this seems like an odd choice.

**Comment by callain** [comment]
@frankieali Hello, can you take @rexxars feedback in consideration ?

**Comment by frankieali** [comment]
I've rebased my PR to include the latest in master. I've added some dev comments to the code and added the [Table of Contents](https://github.com/remarkjs/remark-toc) plugin as a dev-dependency for testing and demo purposes.

The reason for the `root` variable in the global module scope is to cache the root node so it can be walked later when trying to get position attributes from ancestor nodes. Code throughout the `ast-to-react.js` file make references to the `node.position` attribute. Tacking the parent's positioning onto child nodes seems the least obtrusive way to resolve this issue.

**Comment by rexxars** [comment]
Thanks, I'll have a look!

**Comment by simoncorry** [comment]
+1 for implementing this!

**Comment by EricWVGG** [comment]
I gave this PR a try with about five different plugins last night and so far it's working fabulously.

**Comment by cheshire137** [comment]
This pull request is working for me to get the [gemoji](https://github.com/remarkjs/remark-gemoji) and [gemoji-to-emoji](https://github.com/jackycute/remark-gemoji-to-emoji) plugins working!

**Comment by callain** [comment]
Hello @rexxars , hope you're good :) Can you review/merge this PR ? 😄

**Comment by michelhelms-tdg** [comment]
Status? I would love to see that merged and the requested changes seem very minor.

**Comment by rexxars** [comment]
Sorry, just moved to another country, been quite hectic. Will take a look at PRs this weekend.

**Comment by rexxars** [comment]
Thanks so much for this! Released in 4.2.0
